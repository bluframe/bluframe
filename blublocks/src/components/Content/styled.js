// @flow

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Wrapper: StyledComponent<{}, Theme, HTMLElement> = styled.span`
  color: ${({ theme }) => theme.palette.text};
`
