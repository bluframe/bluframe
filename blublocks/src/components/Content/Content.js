// @flow

import type { ComponentProps } from "."
import React from "react"
import ReactMarkdown from "react-markdown"
import { Wrapper } from "./styled"

const Content = ({
  components,
  content,
  rehypePlugins,
  remarkPlugins
}: ComponentProps): React$Node => (
  <Wrapper>
    <ReactMarkdown
      components={components}
      rehypePlugins={rehypePlugins}
      remarkPlugins={remarkPlugins}
    >
      {content}
    </ReactMarkdown>
  </Wrapper>
)

export default Content
