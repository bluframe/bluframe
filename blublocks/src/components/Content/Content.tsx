import type { ComponentProps } from "."
import React from "react"
import ReactMarkdown from "react-markdown"
import { Wrapper } from "./styled"

const Content = ({
  className,
  components,
  content,
  rehypePlugins,
  remarkPlugins
}: ComponentProps) => (
  <Wrapper className={className}>
    <ReactMarkdown
      components={components}
      rehypePlugins={rehypePlugins}
      remarkPlugins={remarkPlugins}
    >
      {content}
    </ReactMarkdown>
  </Wrapper>
)

export default Content
