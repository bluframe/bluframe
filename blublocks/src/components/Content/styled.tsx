import styled from "styled-components"
import { Theme } from "styles/theme"

export const Wrapper = styled.span`
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.main};
`
