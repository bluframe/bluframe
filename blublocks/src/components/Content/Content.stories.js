// @flow

import Content from "components/Content"
import type { ContentProps } from "@bluframe/blublocks"
import React from "react"

const stories = {
  component: Content,
  title: "Components/Content"
}

const Template = (args: ContentProps) => <Content {...args} />

export const Default: any = Template.bind({})
Default.args = {
  content: `# Markdown Content
  Is **the best!!**`
}

export default stories
