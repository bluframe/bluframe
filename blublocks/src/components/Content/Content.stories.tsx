import Content from "components/Content"
import type { Props } from "."
import React from "react"

const stories = {
  component: Content,
  title: "Components/Content"
}

const Template = (args: Props) => <Content {...args} />

export const Default: any = Template.bind({})
Default.args = {
  content: `# Markdown Content
  Is **the best!!**`
}

export default stories
