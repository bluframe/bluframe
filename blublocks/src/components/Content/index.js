// @flow
/* eslint-disable max-lines-per-function */

import type { ContentComponents, ContentProps } from "@bluframe/blublocks"
import { prepareComponent, useTrimText } from "@bluframe/grapple"
import Content from "./Content"
import React from "react"
import strip from "strip-markdown"

type Props = {|
  ...ContentProps
|}

export type ComponentProps = {|
  +className?: string,
  +components: ContentComponents,
  +content: string,
  +rehypePlugins: any[],
  +remarkPlugins: any[]
|}

const usePrepareComponent = ({
  className,
  components: overrideComponents = {},
  content,
  max = Infinity,
  paragraphAs,
  stripMarkdown
}: Props): ComponentProps => {
  const components: ContentComponents = {}

  if (paragraphAs) {
    // eslint-disable-next-line react/display-name, no-unused-vars, id-length
    components.p = ({ node, ...props }) =>
      React.createElement(paragraphAs, props)
  }

  const remarkPlugins = []

  if (stripMarkdown) {
    remarkPlugins.push(strip)
  }

  const rehypePlugins = []

  return {
    className,
    components: {
      ...components,
      ...overrideComponents
    },
    content: useTrimText(content, max),
    rehypePlugins,
    remarkPlugins
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Content
): React$ComponentType<Props>)
