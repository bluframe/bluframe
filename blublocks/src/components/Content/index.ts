/* eslint-disable max-lines-per-function */

import { prepareComponent, useTrimText } from "@bluframe/grapple"
import Content from "./Content"
import React from "react"
import strip from "strip-markdown"

// Find a better solution other than any
export interface ContentComponents {
  // eslint-disable-next-line no-unused-vars
  [tag: string]: (props: any) => any
}

export interface Props {
  className?: string
  components?: ContentComponents
  content: string
  max?: number
  paragraphAs?: string
  stripMarkdown?: boolean
}

export interface ComponentProps
  extends Omit<Props, "max" | "paragraphAs" | "stripMarkdown"> {
  rehypePlugins: any[]
  remarkPlugins: any[]
}

const usePrepareComponent = ({
  className,
  components: overrideComponents = {},
  content,
  max = Infinity,
  paragraphAs,
  stripMarkdown
}: Props): ComponentProps => {
  const components: ContentComponents = {}

  if (paragraphAs) {
    // eslint-disable-next-line react/display-name, no-unused-vars, id-length
    components.p = ({ node, ...props }) =>
      React.createElement(paragraphAs, props)
  }

  const remarkPlugins = []

  if (stripMarkdown) {
    remarkPlugins.push(strip)
  }

  const rehypePlugins: any[] = []

  return {
    className,
    components: {
      ...components,
      ...overrideComponents
    },
    content: useTrimText(content, max),
    rehypePlugins,
    remarkPlugins
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Content
)
