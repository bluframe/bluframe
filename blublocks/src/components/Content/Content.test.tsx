import Content from "./Content"
import React from "react"
import { render } from "tests"

describe("Content", () => {
  it("renders and uses custom components and plugins", () => {
    const mockCodeComponent = jest.fn(() => <code>{"Custom code"}</code>)
    const mockEmComponent = jest.fn(() => <em>{"Custom em"}</em>)

    const props = {
      components: {
        code: mockCodeComponent,
        em: mockEmComponent
      },
      content: `
        ## Markdown Content

        Looks **awesome!!**
      `,
      rehypePlugins: [],
      remarkPlugins: []
    }

    const { container } = render(<Content {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
