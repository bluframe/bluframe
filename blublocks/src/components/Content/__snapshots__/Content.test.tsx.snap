// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content renders and uses custom components and plugins 1`] = `
<span
  class="sc-bdDqHD jkXIsu"
>
  <div>
    
        ## Markdown Content

        Looks **awesome!!**
      
  </div>
</span>
`;
